aten/src/ATen/core/library.cpp

```diff
@@ -62,8 +62,10 @@ void Library::reset() {
 #if defined(TORCH_LIBRARY_THREAD_UNSAFE_LAZY_INIT) && defined(C10_MOBILE)
 namespace detail {
   std::vector<TorchLibraryInit*> torch_library_initializers;
+  std::mutex torch_library_initializer_mutex;
 } // namespace detail
 void initialize_torch_libraries() {
+  const std::lock_guard<std::mutex> lock(detail::torch_library_initializer_mutex);
```

Comment:
I was going to suggest doing this via some sort of call_once equivalent (probably initializing a function-local static), but I see why that's not desirable (we might dlopen some code that adds more initializers). Two requests:
1) please document that fact here so nobody comes along and tries to change it
2) it seems inconsistent that we punt thread safety for the write path but not the read path. Why don't we hold this mutex while we write to torch_library_initializers as well?

Reply:
1. will do
2. the reason is that for the write path we expect library loading to be handled by dyld, so already enforced single threaded before application code runs. In this condition holding a lock is unnecessary overhead. The read path can occur on any thread however, and potentially can race, depending on application architecture.

Reply:
do we have any way to guarantee that nobody will screw that up in the future given that we are now relying on it?